### PR TITLE
Coalesce all memory for checks and reports into shared pointers

### PR DESCRIPTION
**Pull request by duderino**
**What this PR does / why we need it**:

This tries to get all memory associated with mixer requests into shared pointers.   In several places I continue to use raw pointers because I could see that they were only used in the scope of the calling function and I wanted to minimize the size of this already huge PR.

It's necessary for implementing policy check retry on transport error safely (we need to ensure that all the memory needed by a retry lambda is not freed before the lambda is invoked).  

It also fixes a SIGSEGV and memory leak. 

It has a nice side effect of eliminating a few deep copies of protobufs.

Memory Leak Details: 

If the upstream client disconnects during an upstream check to
the policy server, memory is leaked. Example:

```
Leak of 375808 bytes in 367 objects allocated from:
	@ 1fef7cf google::protobuf::internal::ArenaImpl::NewBlock
	@ 1ff011e google::protobuf::internal::ArenaImpl::SerialArena::AllocateAlignedFallback
	@ 1ff09b6 google::protobuf::internal::ArenaImpl::SerialArena::AllocateAligned
	@ 1fefb78 google::protobuf::internal::ArenaImpl::AllocateAligned
	@ 61544f google::protobuf::Arena::CreateMaybeMessage
	@ 55c835 google::protobuf::MessageLite::CreateMaybeMessage
	@ 55b973 istio::mixer::v1::CheckRequest::mutable_attributes
	@ 55a0aa istio::mixerclient::MixerClientImpl::Check
```

SIGSEGV Details: 

If too many requests are sent to the upstream policy server a
circuit breaker may be hit (by default this limits to 1024 concurrent
requests). The MixerClientImpl will free memory prematurely and SIGSEGV
with a stack trace like:

```
[2019-01-26 23:15:47.525][28][debug][pool] [external/envoy/source/common/http/http2/conn_pool.cc:79] max requests overflow
[2019-01-26 23:15:47.525][28][debug][router] [external/envoy/source/common/router/router.cc:532] [C0][S6223237052999230964] upstream reset
[2019-01-26 23:15:47.526][28][debug][http] [external/envoy/source/common/http/async_client_impl.cc:93] async http request response headers (end_stream=true):
':status', '200'
'content-type', 'application/grpc'
'grpc-status', '14'
'grpc-message', 'upstream connect error or disconnect/reset before headers'
'x-envoy-overloaded', 'true'

[2019-01-26 23:15:47.526][28][debug][grpc] [src/envoy/utils/grpc_transport.cc:78] Check failed with code: 14, upstream connect error or disconnect/reset before headers
[2019-01-26 23:15:47.526][28][warning][filter] [src/istio/mixerclient/client_impl.cc:136] Mixer policy check transport error: UNAVAILABLE:upstream connect error or disconnect/reset before headers
[2019-01-26 23:15:47.526][28][debug][filter] [src/istio/mixerclient/client_impl.cc:142] Mixer policy check status: UNAVAILABLE:upstream connect error or disconnect/reset before headers
[2019-01-26 23:15:47.526][28][debug][filter] [src/envoy/http/mixer/filter.cc:152] Called Mixer::Filter : check complete UNAVAILABLE:upstream connect error or disconnect/reset before headers
[2019-01-26 23:15:47.526][28][debug][filter] [src/envoy/http/mixer/filter.cc:132] Called Mixer::Filter : encodeHeaders 3
[2019-01-26 23:15:47.526][28][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:1234] [C2230][S9774838916162700467] encoding headers via codec (end_stream=false):
':status', '503'
'content-length', '69'
'content-type', 'text/plain'
'date', 'Sat, 26 Jan 2019 23:15:46 GMT'
'server', 'envoy'

[2019-01-26 23:15:47.526][28][debug][filter] [src/envoy/http/mixer/filter.cc:207] Called Mixer::Filter : onDestroy state: 3
[2019-01-26 23:15:47.526][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:125] Caught Segmentation fault, suspect faulting address 0x8
[2019-01-26 23:15:47.526][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:94] Backtrace thr<28> obj</usr/lib/x86_64-linux-gnu/libstdc++.so.6> (If unsymbolized, use tools/stack_decode.py):
[2019-01-26 23:15:47.526][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:117] thr<28> #0 0x7ff123114c04 (unknown)
[2019-01-26 23:15:47.526][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] thr<28> obj</home/jblatt/.cache/bazel/_bazel_jblatt/b6258e8f3eabbcfebfab41d31f03811a/sandbox/linux-sandbox/4/execroot/__main__/bazel-out/k8-fastbuild/bin/test/integration/mixer_fault_test.runfiles/__main__/test/integration/mixer_fault_test>
[2019-01-26 23:15:47.553][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #1 0x209e113 google::protobuf::internal::GeneratedMessageReflection::HasBit()
[2019-01-26 23:15:47.571][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #2 0x20923d6 google::protobuf::internal::GeneratedMessageReflection::HasField()
[2019-01-26 23:15:47.587][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #3 0x2037f03 google::protobuf::TextFormat::Printer::PrintField()
[2019-01-26 23:15:47.603][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #4 0x2036d68 google::protobuf::TextFormat::Printer::Print()
[2019-01-26 23:15:47.621][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #5 0x20367e1 google::protobuf::TextFormat::Printer::Print()
[2019-01-26 23:15:47.638][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #6 0x20332f1 google::protobuf::TextFormat::Printer::PrintToString()
[2019-01-26 23:15:47.656][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #7 0x2033144 _ZNK6google8protobuf7Message11DebugStringB5cxx11Ev
[2019-01-26 23:15:47.673][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #8 0x50af1b Envoy::Utils::GrpcTransport<>::GrpcTransport()
[2019-01-26 23:15:47.690][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #9 0x50aaff Envoy::Utils::GrpcTransport<>::GetFunc()::{lambda()#1}::operator()()
[2019-01-26 23:15:47.707][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #10 0x50a857 std::_Function_handler<>::_M_invoke()
[2019-01-26 23:15:47.724][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #11 0x582383 std::function<>::operator()()
[2019-01-26 23:15:47.758][28][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:114] thr<28> #12 0x580b9a istio::mixerclient::MixerClientImpl::Check()
```
**Which issue this PR fixes**

This is another breakout PR of #2107 which is in service of https://github.com/istio/istio/issues/8224.  

I'll follow it with two more PRs (one with more counters, another with retries on policy/quota check transport failures)

This will help find future issues related to [istio/istio#8224](https://github.com/istio/istio/issues/8224)

### CODE REVIEW

File: src/istio/mixerclient/shared_attributes.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  ::istio::mixer::v1::Attributes* attributes() { return attributes_; }
+
+  google::protobuf::Arena& arena() { return arena_; }
+
+ private:
+  google::protobuf::Arena arena_;
+  ::istio::mixer::v1::Attributes* attributes_;
+};
+
+typedef std::shared_ptr<SharedAttributes> SharedAttributesPtr;
```
Comment (PiotrSikora) [inline src/istio/mixerclient/shared_attributes.h]: Nit: `SharedAttributesSharedPtr` (`<class>Ptr` is used for `std::unique_ptr<class>` in Envoy).

Comment (duderino) [reply src/istio/mixerclient/shared_attributes.h]: Will fix.

Anything else I should do or would it be appropriate to give me an LGTM?

**Comment by duderino** [comment]
@JimmyCYJ, Piotr is giving this a review, but I think it'd be really good if you took a look at it too.

**Comment by JimmyCYJ** [comment]
> @JimmyCYJ, Piotr is giving this a review, but I think it'd be really good if you took a look at it too.

Sure, I will review. Thanks.

File: src/istio/mixerclient/shared_attributes.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  SharedAttributes()
+      : attributes_(google::protobuf::Arena::CreateMessage<
+                    ::istio::mixer::v1::Attributes>(&arena_)) {}
+
+  const ::istio::mixer::v1::Attributes* attributes() const {
+    return attributes_;
+  }
+  ::istio::mixer::v1::Attributes* attributes() { return attributes_; }
+
+  google::protobuf::Arena& arena() { return arena_; }
```
Comment (JimmyCYJ) [inline src/istio/mixerclient/shared_attributes.h]: We have tried using arena before. And @mandarjog found there is CPU cost regression. So we revert it here
https://github.com/istio/proxy/pull/1989.
I am not sure if this time the CPU cost issue will come again.

File: src/istio/control/http/attributes_builder.h [hunk 3]
```diff
@@ -52,8 +54,7 @@ class AttributesBuilder {
   // not available.
   void ExtractAuthAttributes(CheckData* check_data);
 
-  // The request context object.
-  RequestContext* request_;
+  istio::mixer::v1::Attributes* attributes_;
```
Comment (PiotrSikora) [inline src/istio/control/http/attributes_builder.h]: Could you add a `TODO` to get rid of this raw pointer?

File: src/istio/mixerclient/client_impl.cc [hunk 4]
```diff
@@ -52,110 +52,122 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
...
-        delete raw_check_result;
-        delete raw_quota_result;
-        delete arena;
-
-        if (utils::InvalidDictionaryStatus(status)) {
-          compressor_.ShrinkGlobalDictionary();
-        }
-      });
+  return transport(context->request(), context->response(),
+                   [this, context = context, on_done](const Status &status) {
```
Comment (silentdai) [inline src/istio/mixerclient/client_impl.cc]: is this lambda invoked immediately in transport, or scheduled later? If latter, `*this` may be destructed.

File: src/istio/mixerclient/shared_attributes.h (hunk 2 above)
Comment (duderino) [reply src/istio/mixerclient/shared_attributes.h]: Recapping a gchat conversation here:  we agreed the Arena should be evaluated but since this was just moved here and not added we want to minimize changes to this already overly-large PR and address in a subsequent PR

**Comment by duderino** [comment]
@JimmyCYJ I already ran this against the mixer client tests in istio/istio.  This PR will have to be merged after istio/istio pulls in the new istio/proxy:  https://github.com/istio/istio/pull/11591

File: src/istio/mixerclient/check_context.h [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    final_status_ = status;
+  }
+
+  //
+  // CheckResponseInfo (exposed to the top-level filter)
+  //
+
+  virtual const google::protobuf::util::Status& status() const override {
```
Comment (PiotrSikora) [inline src/istio/mixerclient/check_context.h]: Nit: this doesn't need to be `virtual` anymore.

File: src/istio/mixerclient/check_context.h [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    return final_status_;
+  }
+
+  virtual const istio::mixer::v1::RouteDirective& routeDirective()
```
Comment (PiotrSikora) [inline src/istio/mixerclient/check_context.h]: Nit: this doesn't need to be `virtual` anymore.

File: include/istio/mixerclient/check_response.h [hunk 7]
```diff
@@ -22,21 +22,15 @@
...
-  bool is_check_cache_hit{false};
+// The CheckResponseInfo exposes policy and quota check details to the check
+// callbacks.
+class CheckResponseInfo {
+ public:
+  virtual ~CheckResponseInfo(){};
 
-  // Whether this quota response is from cache.
-  bool is_quota_cache_hit{false};
+  virtual const ::google::protobuf::util::Status& status() const = 0;
```
Comment (PiotrSikora) [inline include/istio/mixerclient/check_response.h]: Nit: Envoy uses `PURE` macro for `= 0`;

File: include/istio/mixerclient/check_response.h [hunk 8]
```diff
@@ -22,21 +22,15 @@
...
 
-  // The check and quota response status.
-  ::google::protobuf::util::Status response_status{
-      ::google::protobuf::util::Status::UNKNOWN};
-
-  // Routing directive (applicable if the status is OK)
-  ::istio::mixer::v1::RouteDirective route_directive{
-      ::istio::mixer::v1::RouteDirective::default_instance()};
+  virtual const ::istio::mixer::v1::RouteDirective& routeDirective() const = 0;
```
Comment (PiotrSikora) [inline include/istio/mixerclient/check_response.h]: Nit: Envoy uses `PURE` macro for `= 0`;

File: src/istio/control/client_context_base.cc [hunk 9]
```diff
@@ -79,37 +79,24 @@ ClientContextBase::ClientContextBase(const TransportConfig& config,
...
-    utils::AttributesBuilder builder(request->attributes);
-    builder.AddBool(utils::AttributeName::kCheckCacheHit,
-                    check_response_info.is_check_cache_hit);
-    builder.AddBool(utils::AttributeName::kQuotaCacheHit,
-                    check_response_info.is_quota_cache_hit);
-    on_done(check_response_info);
-  };
-
+CancelFunc ClientContextBase::SendCheck(
+    TransportCheckFunc transport, CheckDoneFunc on_done,
```
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/control/client_context_base.cc]: Yes, this code has many layers and I don't think they are all necessary.   A big refactoring is in order but I think everyone wants to minimize changes here so not this pass :)

File: src/istio/mixerclient/check_context.h [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+  void compressRequest(const AttributeCompressor& compressor,
+                       const std::string& deduplication_id) {
+    compressor.Compress(*shared_attributes_->attributes(),
+                        allocRequestOnce()->mutable_attributes());
+    request_->set_global_word_count(compressor.global_word_count());
+    request_->set_deduplication_id(deduplication_id);
+  }
+
+  bool failOpen() const { return fail_open_; }
```
Comment (PiotrSikora) [inline src/istio/mixerclient/check_context.h]: This seem to be inheriting value from `ClientContextBase::NetworkFailOpen()`, but uses a different name. Could you change one or the other to keep things consistent?

File: src/istio/mixerclient/check_context.h [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+  istio::mixer::v1::CheckRequest* allocRequestOnce() {
+    if (!request_) {
+      request_ = google::protobuf::Arena::CreateMessage<
+          istio::mixer::v1::CheckRequest>(&shared_attributes_->arena());
+    }
+
+    return request_;
+  }
+
+  istio::mixerclient::SharedAttributesSharedPtr shared_attributes_;
```
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/mixerclient/check_context.h]: Good catch.  This was a mistake that I fixed up in a later PR in the series.  The whole section can be private.  Will fix here too

File: src/istio/mixerclient/client_impl.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/mixerclient/client_impl.cc]: Yeah, I guess this is unnecessary.  I think I did it to emphasize how important it is to inc the ref count on context.  That did make it catch your eye ;)

Comment (duderino) [reply src/istio/mixerclient/client_impl.cc]: Lamda may be invoked later.   The this pointer refers to the MixerClientImpl object which has a much longer lifetime than any individual transaction (in this case the CheckContext has transaction lifespan).    

But fundamentally I'm relying on no transactions completing after the MixerClientImpl is destroyed like the previous code and it's possible there may still be issues there.

File: src/istio/mixerclient/client_impl.cc [hunk 12]
```diff
@@ -52,110 +52,122 @@ MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
...
+
+                     if (utils::InvalidDictionaryStatus(status)) {
+                       compressor_.ShrinkGlobalDictionary();
+                     }
+                   });
 }
 
-void MixerClientImpl::Report(const Attributes &attributes) {
-  report_batch_->Report(attributes);
+void MixerClientImpl::Report(const SharedAttributesSharedPtr &context) {
```
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/mixerclient/client_impl.cc]: wtf?  thanks :)

File: src/istio/mixerclient/shared_attributes.h [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+ private:
+  google::protobuf::Arena arena_;
```
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/mixerclient/shared_attributes.h]: yes, the above is true

File: include/istio/mixerclient/check_response.h (hunk 7 above)
Comment (PiotrSikora) [inline include/istio/mixerclient/check_response.h]: Why rename `response_status()` to `status()`? This resulted in a lot of seemingly unnecessary changes in this PR.

File: src/istio/control/http/attributes_builder.h (hunk 3 above)
Comment (duderino) [reply src/istio/control/http/attributes_builder.h]: happily

File: src/istio/mixerclient/check_context.h (hunk 5 above)
Comment (duderino) [reply src/istio/mixerclient/check_context.h]: for my own education:  because of the override keyword, or because it inherits from a class that already declared this virtual?

File: include/istio/mixerclient/check_response.h (hunk 7 above)
Comment (duderino) [reply include/istio/mixerclient/check_response.h]: Yeah, I think I'd have to define it here since we are in the mixerclient tree and no direct dependencies on Envoy are allowed.

File: src/istio/mixerclient/check_context.h (hunk 10 above)
Comment (duderino) [reply src/istio/mixerclient/check_context.h]: will change to networkFailOpen()

File: include/istio/mixerclient/check_response.h (hunk 7 above)
Comment (duderino) [reply include/istio/mixerclient/check_response.h]: we discussed and decided it would have been good to do this in the first place, but is now unnecessary

**Comment by JimmyCYJ** [comment]
> @JimmyCYJ I already ran this against the mixer client tests in istio/istio. This PR will have to be merged after istio/istio pulls in the new istio/proxy: [istio/istio#11591](https://github.com/istio/istio/pull/11591)

Thanks. 

/lgtm

File: src/istio/mixerclient/check_context.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/mixerclient/check_context.h]: fixed

File: include/istio/mixerclient/check_response.h (hunk 7 above)
Comment (duderino) [reply include/istio/mixerclient/check_response.h]: We discussed and decided that since istio/mixerclient can't depend on Envoy sources, it's ok to not make this change

File: src/istio/mixerclient/check_context.h [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+    final_status_ = status;
+  }
+
+  //
+  // CheckResponseInfo (exposed to the top-level filter)
+  //
+
+  virtual const google::protobuf::util::Status& status() const override {
+    return final_status_;
```
[the remark this answers is not in the document]
Comment (duderino) [reply src/istio/mixerclient/check_context.h]: While both setFinalStatus() and status() may be invoked by different threads, they aren't invoked concurrently
